flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -318,6 +392,143 @@ private void outputCompletedElement() {
+        }
+
+        private void cleanupLastRetryInMailbox() {
+            if (retryInFlight.compareAndSet(true, false)) {
+                assert incompleteDelayRetryHandlers.contains(this);
+                // remove from delayed retry queue
+                incompleteDelayRetryHandlers.remove(this);
+                delayedRetryTimer = null;
+            }
+        }
+
+        private boolean ifRetryOrCompleted(Collection<OUT> results, Throwable error) {
+            if (delayedRetryAvailable.get() && resultHandler.inputRecord.isRecord()) {
+                boolean satisfy = false;
+                if (null != results && null != retryResultPredicate) {
+                    satisfy = (satisfy || retryResultPredicate.test(results));
+                }
+                if (null != error && null != retryExceptionPredicate) {
+                    satisfy = (satisfy || retryExceptionPredicate.test(error));
+                }
+                if (satisfy) {
+                    if (asyncRetryStrategy.canRetry(getCurrentAttempts())) {
+                        long nextBackoffTimeMillis =
+                                asyncRetryStrategy.getBackoffTimeMillis(getCurrentAttempts());
+                        setBackoffTimeMillis(nextBackoffTimeMillis);
+                        if (delayedRetryAvailable.get()) {
+                            // timer thread will finally dispatch the task to mailbox executor
+                            mailboxExecutor.submit(
+                                    () -> trySubmitRetryInMailboxOrComplete(this, results, error),
+                                    "delayed retry or give up retry");
+                            return true;
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        private void trySubmitRetryInMailboxOrComplete(
+                RetryableResultHandlerDelegator resultHandlerDelegator,
+                Collection<OUT> results,
+                Throwable error) {
+            if (delayedRetryAvailable.get()) {
+                incompleteDelayRetryHandlers.add(resultHandlerDelegator);
+                retryInFlight.set(true);
+                final long delayedRetry =
+                        resultHandlerDelegator.getBackoffTimeMillis()
+                                + getProcessingTimeService().getCurrentProcessingTime();
+                delayedRetryTimer =
+                        processingTimeService.registerTimer(
+                                delayedRetry, timestamp -> doRetry(resultHandlerDelegator));
+            } else {
+                // give up retry and complete immediately
+                if (null != results) {
+                    resultHandler.complete(results);
+                } else {
+                    resultHandler.completeExceptionally(error);
+                }
+            }
+        }
+
+        public long getBackoffTimeMillis() {
```

Comment:
We might could remove some of the getter / setter since they are only used in this class?